src/base64.cr

```diff
@@ -156,36 +156,196 @@ module Base64
     strict_encode_to_io_internal(data, io, CHARS_SAFE, pad: true)
   end
 
-  # Returns the base64-decoded version of *data* as a `Bytes`.
+  # Writes the base64-decoded version of `from` into `to`,
+  # and returns the amount of bytes written.
+  #
+  # If the `to` buffer is not big enough to store the resulting bytes,
+  # or if the input is invalid base64, the method raises a `Base64::Error`.
+  #
+  # This will decode either the normal or urlsafe alphabets.
+  def decode(from : Bytes, to : Bytes) : Int32
+    total_written_bytes : Int32 = 0
+    total_read_bytes : Int32 = 0
+
+    from_ptr = from.to_unsafe
+    from_size = from.size
+    to_ptr = to.to_unsafe
+    to_size = to.size
+
+    read_bytes, written_bytes = decode_buffer_regular(from_ptr, from_size, to_ptr, to_size)
+    total_written_bytes += written_bytes
+    total_read_bytes += read_bytes
+    from_ptr += read_bytes
+    from_size &-= read_bytes
+    to_ptr += written_bytes
+    to_size &-= written_bytes
+
+    if from_size > 0
+      read_bytes, written_bytes = decode_buffer_final(from_ptr, from_size, to_ptr, to_size)
+      total_written_bytes += written_bytes
+      total_read_bytes += read_bytes
+
+      from_ptr += read_bytes
+      from_size &-= read_bytes
+      to_ptr += written_bytes
+      to_size &-= written_bytes
+
+      raise Base64::Error.new("Expected EOF but found #{from_size} additional bytes") if from_size > 0
+    end
+
+    total_written_bytes
+  end
+
+  # Returns the base64-decoded version of `data` as a `Bytes`.
+  #
+  # Raises `Base64::Error` if the input is invalid base64.
+  #
   # This will decode either the normal or urlsafe alphabets.
   def decode(data) : Bytes
-    slice = data.to_slice
-    buf = Pointer(UInt8).malloc(decode_size(slice.size))
-    appender = buf.appender
-    from_base64(slice) { |byte| appender << byte }
-    Slice.new(buf, appender.size.to_i32)
+    data = data.to_slice
+    buf_size = decode_size(data.size)
+    buf = Pointer(UInt8).malloc(buf_size)
+
+    bytes_written = decode(from: data, to: Bytes.new(buf, buf_size))
+    Bytes.new(buf, bytes_written)
```

Comment:
This can be shortened and simplified:

```suggestion
    buffer = Bytes.new(buf_size)
    bytes_written = decode(from: data, to: buffer)
    buffer[...bytes_written]
```